rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,541 @@
+  instance.** Allowing many joins on one token creates unnecessary backend
+  contention and - depending on implementation - creates severe traceability
+  problems.
+
+  However, we can greatly improve today's automation story. Secret fulfillment
+  can take place server side to reduce the number of resources to generate in
+  Terraform, and `ProvisionToken` resources can be made reusable to fully enable
+  IaC workflows.
+
+- **Bot credentials can be long-lived, but their *trust* must be controlled and
+  renewed.** We can't allow any secret values to either remain both valid and
+  useful for a long time, or allow currently valid identities to generate
+  credentials that can be used for future, unchecked extension of access.
+
+  However, we can create a new state where an identity is technically valid
+  indefinitely, but useless unless explicitly allowed by the server. Existing
+  controls like the generation counter can still effectively prevent unintended
+  reuse of the bot identity.
+
+### Challenge-Response Joining
+
+TODO: Consider alternative join method names?
+
+We believe a new join method, `challenge`, can meet our needs and provide
+significantly more flexibility than today's `token` join method. This works by -
+in a sense - inverting the token joining procedure: bots generate an ED25519
+keypair, and the public key is copied to the server. The public key can be
+copied out-of-band, or bots can provide their public key on first join using a
+one-time use shared secret to authenticate the exchange, much like today's
+`token` method.
+
+Once the public key has been shared, bots may then join by requesting a
+challenge from the Teleport Auth service and completing it by signing it with
+their private key. If successful, the bot is issued a renewable identity just as
+`token`-joined bots are today, and the bot will actively renew this identity for
+as long as possible, or until its backing token expires.
+
+If the identity renewal fails at any point, bots may attempt to reauthenticate,
+and the Auth service can use predefined per-bot rules to decide if this specific
+bot is allowed to rejoin, including a rejoin counter and expiration date. If a
+rejoin is rejected, the bot's identity does not necessarily remain invalid: if
+server-side rules are adjusted, for example by increasing the token's rejoin
+limit, it can then rejoin without any client-side reconfiguration.
+
+This has several important differences to existing join methods:
+
+- Onboarding secrets are optional, and the secret exchange process may be
+  skipped if the `ProvisionToken` is configured with a public key directly.
+  Otherwise, joining bots authenticate with an onboarding secret to
+  automatically share their public key with the server.
+
+- When joining or rejoining, Teleport issues a challenge that the client must
+  solve. This is similar to TPM joining today, but backed by a local keypair
+  rather than (necessarily) a hardware token.
```

Comment:
_Could_ it also work with a TPM though?

Reply:
There's a short section on this, but if it supports PKCS#11 key storage then yes, we'd like to support that by making use of our existing HSM library support. I intend to keep the storage backend extensible to make supporting this straightforward. I could also see a macOS-specific backend, for example.

Reply:
I've added more details on this and explicitly called out support for TPMs/secure enclave.